src/plugins/discover/public/utils/get_sharing_data.ts

```diff
@@ -96,11 +96,13 @@ export async function getSharingData(
        * Discover does not set fields, since having all fields is needed for the UI.
        */
       const useFieldsApi = !config.get(SEARCH_FIELDS_FROM_SOURCE);
-      if (useFieldsApi && columns.length) {
-        searchSource.setField(
-          'fields',
-          columns.map((field) => ({ field, include_unmapped: 'true' }))
-        );
+      if (useFieldsApi) {
+        searchSource.removeField('fieldsFromSource');
+        const fields = columns.length
+          ? columns.map((field) => ({ field, include_unmapped: 'true' }))
+          : [{ field: '*', include_unmapped: 'true' }];
```

Comment:
Without this change, the CSV export broke. 
Why?  `[{ field: '*', include_unmapped: 'true' }]` is necessary to export all the fields given useFieldsApi is `true` (Which is the default)
Bevor the refactoring of this PR this was added by the code executed in the hook. So it was added before , and at a different part of the code, which made it hard to track & debug

This is now slightly better, more decoupled, independent from previous searchSource mutations